[Feature] Linear onboarding with custom backend - Milestone 1

mejdoo opened the pull request:
This PR introduces the new feature Linear onboarding with custom backend. 

## What's new in this PR?

-  New welcome screen in Kotlin which can be also accessed from "New Team or Account" in the profile.
-  Removed login button from the previous landing screen. Can be accessed from the button "CREATE AN ACCOUNT" in the welcome screen 
- Removed SSO Login option from login screen and move it the new welcome screen (Button : Enterprise Log IN)
- Updated title and message for SSO Login Dialog




#### APK
[Download build #1063](http://10.10.124.11:8080/job/Pull%20Request%20Builder/1063/artifact/build/artifact/wire-dev-PR2567-1063.apk)
[Download build #1064](http://10.10.124.11:8080/job/Pull%20Request%20Builder/1064/artifact/build/artifact/wire-dev-PR2567-1064.apk)
[Download build #1066](http://10.10.124.11:8080/job/Pull%20Request%20Builder/1066/artifact/build/artifact/wire-dev-PR2567-1066.apk)
[Download build #1070](http://10.10.124.11:8080/job/Pull%20Request%20Builder/1070/artifact/build/artifact/wire-dev-PR2567-1070.apk)
[Download build #1074](http://10.10.124.11:8080/job/Pull%20Request%20Builder/1074/artifact/build/artifact/wire-dev-PR2567-1074.apk)

mejdoo commented on app/src/main/res-v2/layout/fragment_welcome.xml:
```diff
@@ -0,0 +1,135 @@
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+-->
+<androidx.constraintlayout.widget.ConstraintLayout xmlns:android="http://schemas.android.com/apk/res/android"
```
I used here constraint layout with percentages and guidelines. Using this approach we can support multiple type of screens with a single layout : small - normal....

BradleyWire replied:
Do we need this many guidelines? We can still get a reactive layout by constraining some components to the parent itself and then using guidelines as a leverage for positioning?

mejdoo replied:
There are multiple approaches to design responsive layouts with ContraintLayout. In my previous job, I tested this approach (multiple guidelines with percentages) in many smartphones with screen sizes from 4 to 6.5 inches. If we use ContraintLayout with paddings and margins, we can discover later malformed design in some screens with small sizes.

mejdoo replied:
We can open a discussion in the next collective if you want to unify our approach how to build response layouts with ContraintLayout.

mejdoo commented on app/src/main/res-v2/layout/fragment_welcome.xml:
```diff
@@ -0,0 +1,135 @@
+        app:layout_constraintEnd_toEndOf="parent"
+        app:layout_constraintTop_toBottomOf="@+id/guideline1" />
+
+    <androidx.appcompat.widget.AppCompatTextView
```
I didn't use any styling here. We have to ask Victor to provide for as a Design Kit to unify our layout styles for the new codebase.

BradleyWire commented on app/src/main/kotlin/com/waz/zclient/auth/WelcomeFragment.kt:
```diff
@@ -0,0 +1,57 @@
+        super.onViewCreated(view, savedInstanceState)
+        create_account_button.setOnClickListener { createAccount() }
+        login_button.setOnClickListener { login() }
+        enterprise_login_button.setOnClickListener {ssoLogin() }
```
formatting

BradleyWire commented on app/src/main/kotlin/com/waz/zclient/auth/WelcomeFragment.kt:
```diff
@@ -0,0 +1,57 @@
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+        create_account_button.setOnClickListener { createAccount() }
```
We can change these id references now to adhere to the style guide seeing as we have agreed on a way forward

mejdoo replied:
done

BradleyWire commented on app/src/main/res/values/strings.xml:
```diff
@@ -1453,4 +1453,8 @@
     <string name="shortcut_new_message_label">New Message</string>
 
+    <!-- SSO Login -->
+    <string name="sso_login_dialog_title">Enterprise log in</string>
```
If we're adding strings to this file, do we need the one in res-v2?

BradleyWire commented on app/src/main/scala/com/waz/zclient/appentry/AppEntryActivity.scala:
```diff
@@ -279,8 +287,7 @@ class AppEntryActivity extends BaseActivity {
       progressView.show(LoadingIndicatorView.SpinnerWithDimmedBackground(), darkTheme = true)
-    else
-      progressView.hide()
+    else if(progressView != null)  progressView.hide()
```
Formatting

BradleyWire replied:
formatting

BradleyWire commented on app/src/main/scala/com/waz/zclient/appentry/SSOFragment.scala:
```diff
@@ -44,8 +43,8 @@ trait SSOFragment extends FragmentHelper with DerivedLogTag {
     override def onDialogEvent(event: Event): Unit = event match {
-      case OnNegativeBtn        => verbose(l"Negative")
-      case OnPositiveBtn(input) => verifyInput(input)
+      case OnNegativeBtn          =>  getActivity.finish()
```
getActivity can be null in some instances, you might be better off calling requireActivity

BradleyWire commented on app/src/main/scala/com/waz/zclient/appentry/TransparentSSOActivity.scala:
```diff
@@ -0,0 +1,41 @@
+    else setTheme(R.style.Theme_Light_Transparent)
+
+    setContentView(R.layout.transparent_layout)
+    replaceFragment()
```
replaceFragment seems too ambiguous

BradleyWire commented on app/src/main/scala/com/waz/zclient/appentry/fragments/SignInFragment.scala:
```diff
@@ -155,7 +155,7 @@ class SignInFragment
     phoneButton.foreach(_.setVisible(!onlyLogin))
 
-    companyLoginButton.foreach(_.setVisible(BuildConfig.ALLOW_SSO))
+    //companyLoginButton.foreach(_.setVisible(BuildConfig.ALLOW_SSO))
```
If we don't need this, could we remove it?

BradleyWire replied:
If we don't need this can it be removed?

BradleyWire commented on app/src/main/kotlin/com/waz/zclient/auth/WelcomeFragment.kt:
```diff
@@ -0,0 +1,57 @@
+        enterprise_login_button.setOnClickListener {ssoLogin() }
+    }
+
+    private fun createAccount() {
```
This method indicates it'll do something else than starting a new screen, same with other methods

mejdoo replied:
changed it to startCreateAccountFlow()

BradleyWire commented on app/src/main/res-v2/drawable/blue_rounded_button.xml:
```diff
@@ -0,0 +1,26 @@
+<shape xmlns:android="http://schemas.android.com/apk/res/android"
+    android:shape="rectangle">
+    <solid android:color="@color/curious_blue" />
+    <corners android:radius="6dp" />
```
Avoid hardcoded dimens, same in other custom drawables

BradleyWire commented on app/src/main/res-v2/layout/fragment_welcome.xml:
```diff
@@ -0,0 +1,135 @@
+        android:layout_height="0dp"
+        android:background="@drawable/light_gray_blue_rounded_button"
+        android:text="@string/app_entry_login"
+        android:textSize="12sp"
```
If you're still waiting on a design it's probably good to still assign these to dimens and then reassign later if something changes?

BradleyWire commented on app/src/main/res-v2/values/strings.xml:
```diff
@@ -0,0 +1,27 @@
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+-->
+<resources xmlns:tools="http://schemas.android.com/tools">
```
What's the criteria for us adding strings to this file?

mejdoo replied:
I created this file for the new string ids used only by the Kotlin code. If you don't like it, I can remove it and move the fields to the old string.xml file.

gizemb replied:
I think it's safe to move these to strings.xml file

mejdoo replied:
ok done

BradleyWire commented on app/src/main/scala/com/waz/zclient/appentry/TransparentSSOActivity.scala:
```diff
@@ -0,0 +1,41 @@
+import com.waz.zclient._
+
+class TransparentSSOActivity
+  extends BaseActivity {
```
Does this need to be line-wrapped?

BradleyWire commented on app/src/main/scala/com/waz/zclient/appentry/AppEntryActivity.scala:
```diff
@@ -127,14 +129,20 @@ class AppEntryActivity extends BaseActivity {
     closeButton.onClick(abortAddAccount())
 
-    showFragment()
+    getIntent.getStringExtra(WelcomeFragment.BUNDLE_KEY_FRAGMENT_TO_START) match {
```
I'm not sure on the details, but, couldn't we start this activity with a factory method and have the constants live in this class? I don't think we should have a dependency on the caller of this class

gizemb replied:
we don't need to pass a bundle here, we can check getIntent().getAction() and see whether it's our custom action.

mejdoo replied:
done

BradleyWire commented on app/src/main/scala/com/waz/zclient/appentry/TransparentSSOFragment.scala:
```diff
@@ -0,0 +1,38 @@
+class TransparentSSOFragment extends SSOFragment {
+
+  override def onCreateView(inflater: LayoutInflater, container: ViewGroup, savedInstanceState: Bundle): View =
+    inflater.inflate(R.layout.transparent_layout, container, false)
```
style guide says this should be called fragment_transparent_sso

BradleyWire commented on app/src/main/res-v2/values/colors.xml:
```diff
@@ -0,0 +1,26 @@
+<?xml version="1.0" encoding="utf-8"?><!--
+
+    Wire
+    Copyright (C) 2018 Wire Swiss GmbH
```
This can be updated to 2020

gizemb commented on app/src/main/scala/com/waz/zclient/appentry/SSOFragment.scala:
```diff
@@ -44,8 +43,8 @@ trait SSOFragment extends FragmentHelper with DerivedLogTag {
     override def onDialogEvent(event: Event): Unit = event match {
-      case OnNegativeBtn        => verbose(l"Negative")
-      case OnPositiveBtn(input) => verifyInput(input)
+      case OnNegativeBtn          =>  getActivity.finish()
```
discussed in private. will be changed anyway

mejdoo commented on app/src/main/scala/com/waz/zclient/appentry/AppEntryActivity.scala:
```diff
@@ -279,8 +287,7 @@ class AppEntryActivity extends BaseActivity {
       progressView.show(LoadingIndicatorView.SpinnerWithDimmedBackground(), darkTheme = true)
-    else
-      progressView.hide()
+    else if(progressView != null)  progressView.hide()
```
App crashes sometimes with NullPointerException here. I don't have a valid scenario now.

makingthematrix replied:
I suspect `NullPointerException` may be thrown when `enableProgress(false)` is called from `onCreate` (line 136). It might be too early. 
You may try to move that call to  the inside of `deepLinkService.deepLink.collect { case Some(result) => result } onUi { ... }` as this what we use it for here : We call `enableProgress(true)` when switching to another backend and  back to `false` afterwards. So in all other cases, you need to call `enableProgress(false)`.
I know it's a bit messy, but this progress indicator is old code and I don't want to spend much time on it.

By the way, if you need to check for null in Scala, wrap the variable in `Option(...)`. Then you can use `.map`, `.foreach`, `.getOrElse`, or just check `.isDefined`.

mejdoo commented on app/src/main/res-v2/values/colors.xml:
```diff
@@ -0,0 +1,26 @@
+-->
+<resources>
+    <!-- Color names are Inspired from here is http://chir.ag/projects/name-that-color/#566433 -->
+    <color name="outer_space">#323639</color>
```
The color names are inspired from here http://chir.ag/projects/name-that-color/#566433 to avoid redundant colors with different names. We can bring naming conventions for the colors in the next collective.

gizemb replied:
I'd prefer if it included the color name like blue, etc.

mejdoo replied:
If we have multiple blue colors how could we to name them? I have added a topic "Color naming conventions" to the next Android collective.

gizemb replied:
i mean, not blue only, something_blue. i don't know, we can discuss in the meeting.

mejdoo commented on app/src/main/res/values/themes.xml:
```diff
@@ -406,4 +406,23 @@
     </style>
 
+
+    <style name="Theme.Light.Transparent" parent="Theme.Light">
```
No. Theme.Light.Transparent inherits from Theme.Light and Theme.Dark.Transparent from Theme.Dark but the additional attributes are similar. I used this approach to ensure that the selected theme by the user will be applied to the SOO AlertDialog. If I apply directly Transparent Theme to TransparentSSOActivity, the AlertDialog will look very weird.

gizemb replied:
If you say Theme.Light.Transparent, that means you're creating a theme called Transparent, that inherits from Theme.Light (the dots create the inheritance), so you don't have to explicitly specify parent=Theme.Light. that's what I meant above.

mejdoo replied:
I didn't get it in your first comment. Fixed !

makingthematrix commented on app/src/main/scala/com/waz/zclient/InputDialog.scala:
```diff
@@ -106,14 +106,16 @@ class InputDialog extends DialogFragment with FragmentHelper {
       })
       .setNegativeButton(getArguments.getInt(NegativeBtn), new DialogInterface.OnClickListener {
-        def onClick(dialog: DialogInterface, which: Int): Unit =
+        def onClick(dialog: DialogInterface, which: Int): Unit = {
```
No `{ ... }` are needed for expressions ;)

mejdoo commented on app/src/main/kotlin/com/waz/zclient/auth/WelcomeFragment.kt:
```diff
@@ -0,0 +1,57 @@
+        private val ACTION_SSO_LOGIN = Config.applicationId() + ".SSO_LOGIN_ACTION"
+        const val BUNDLE_KEY_FRAGMENT_TO_START = "fragmentToStart"
+        const val BUNDLE_VALUE_LOGIN_FRAGMENT = "LoginFragment"
+        const val BUNDLE_VALUE_SSO_FRAGMENT = "ssoFragment"
```
fixed

mejdoo commented on app/src/main/scala/com/waz/zclient/appentry/TransparentSSOFragment.scala:
```diff
@@ -0,0 +1,38 @@
+  override def onViewCreated(view: View, savedInstanceState: Bundle): Unit =
+    extractTokenAndShowSSODialog(showIfNoToken = true)
+
+  override protected def activity: AppEntryActivity = getActivity.asInstanceOf[AppEntryActivity]
```
I will not get rid of this now to avoid refactoring a lot of Scala code in different activities. I fixed the bug related to the cancelation of SSO AlertDialog by adding this abstract method to SSOFragment. 
> protected def isParentActivityTransparent: Boolean
Then I override it in the fragments which inherit from SSOFragment.

gizemb commented on app/src/main/AndroidManifest.xml:
```diff
@@ -228,6 +244,18 @@
+
+        <activity
+            android:name=".appentry.TransparentSSOActivity"
+            android:theme="@android:style/Theme.Translucent.NoTitleBar"
```
Should this activity be opened from other apps/deeplink? If not, we should add exported=false

mejdoo replied:
done

gizemb commented on app/src/main/AndroidManifest.xml:
```diff
@@ -214,7 +219,18 @@
+            android:exported="false"
+            >
+            <intent-filter>
+                <action android:name="${applicationId}.CREATE_ACCOUNT_ACTION" />
```
did you try that?

mejdoo replied:
yes, it doesn't work as ${applicationId} is not recognised as an application iD when we put it inside strings.xml. It works only when we put it in AndroidManifest.

gizemb replied:
do we need applicationId at all? it can be any string i guess 🤔

gizemb commented on app/src/main/kotlin/com/waz/zclient/core/config/Config.kt:
```diff
@@ -3,10 +3,13 @@ package com.waz.zclient.core.config
     fun versionName() = BuildConfig.VERSION_NAME
     fun websiteUrl() = BuildConfig.WEBSITE_URL
     fun accountsUrl() = BuildConfig.ACCOUNTS_URL
+    fun allowSso() = BuildConfig.ALLOW_SSO
```
why is this determined by BuildConfig?

android10 replied:
it is not clear for me either.

mejdoo replied:
Check default.json file, all the attributes there are listed as a to buildConfigField in app.gradle. This is a legacy configuration. we have to find a way how to get rid of that.

gizemb replied:
I still don't get why we're dependin to default.json file on this. Shouldn't this feature be enabled all the time?

gizemb replied:
ok so it wasn't always allowed before, so it makes sense to keep the same logic

gizemb commented on app/src/main/scala/com/waz/zclient/appentry/FragmentTransactionHelper.scala:
```diff
@@ -0,0 +1,24 @@
+import androidx.fragment.app.{Fragment, FragmentActivity, FragmentTransaction}
+import com.waz.zclient.R
+
+class FragmentTransactionHelper() {
```
please change the name to indicate that this is only a helper for SSO flow.

android10 replied:
Maybe something like TransactionHandler or something without the 'helper' word since the responsibility is not clear

mejdoo replied:
@gizemb I don't think this is not only for SSO flow. Based on your commit,  the method **showFragment in AppEntryActivity** is called from multiple places and it is using this helper.

mejdoo replied:
I change it now to TransactionHandler

android10 commented on app/src/main/kotlin/com/waz/zclient/auth/WelcomeFragment.kt:
```diff
@@ -0,0 +1,65 @@
+        startActivity(loginIntent)
+    }
+
+    private fun startSsoLoginFlow() {
```
This can be inline and simplified to:
```kotlin
private fun startSsoLoginFlow() = startActivity(Intent(ACTION_SSO_LOGIN))
```

android10 commented on app/src/main/kotlin/com/waz/zclient/auth/WelcomeFragment.kt:
```diff
@@ -0,0 +1,65 @@
+        startActivity(createAccountIntent)
+    }
+
+    private fun startLoginFlow() {
```
this can also be more concise. It is also a matter of taste. 
```kotlin
private fun startLoginFlow() =
        startActivity(Intent(ACTION_LOGIN).apply {
            putExtra(BUNDLE_KEY_FRAGMENT_TO_START, BUNDLE_VALUE_LOGIN_FRAGMENT)
        })
```

mejdoo replied:
Done. I added an extension startActivityWithAction to reduce redundancy

android10 commented on app/src/main/kotlin/com/waz/zclient/auth/WelcomeFragment.kt:
```diff
@@ -0,0 +1,65 @@
+        configureSSOButtonVisibility()
+    }
+
+    private fun startCreateAccountFlow() {
```
This can be inline:
```kotlin
    private fun startCreateAccountFlow() =
        startActivity(Intent().apply { action = ACTION_CREATE_ACCOUNT })
```

android10 commented on app/src/main/kotlin/com/waz/zclient/auth/WelcomeFragment.kt:
```diff
@@ -0,0 +1,65 @@
+        }
+    }
+
+    companion object {
```
Maybe some spaces to separate the concerns?
```kotlin
companion object {
        fun newInstance() = WelcomeFragment()
        
        private val ACTION_LOGIN = Config.applicationId() + ".LOGIN_ACTION"
        private val ACTION_CREATE_ACCOUNT = Config.applicationId() + ".CREATE_ACCOUNT_ACTION"
        private val ACTION_SSO_LOGIN = Config.applicationId() + ".SSO_LOGIN_ACTION"
        
        const val BUNDLE_KEY_FRAGMENT_TO_START = "fragmentToStart"
        const val BUNDLE_VALUE_LOGIN_FRAGMENT = "LoginFragment"
    }
```

mejdoo commented on app/src/main/scala/com/waz/zclient/LaunchActivity.scala:
```diff
@@ -96,7 +96,7 @@ class LaunchActivity extends AppCompatActivity with ActivityHelper with DerivedL
 
   private def startSignUp() = {
-    startActivity(new Intent(this, classOf[AppEntryActivity]))
+    startActivity(new Intent(this, classOf[AuthenticationActivity]))
```
done

mejdoo commented:
@android10 You are right, the PR size is over the limit, it would be divided in 3 or 4 PR probably like the bullet points in the description.

For the unit tests, I think there is no business logic to test for this PR. Basically I touched only the UI side: I created a new welcome screen, added an interaction between the Kotlin activities and the Scala activities via custom Intents and removed some buttons in Scala side. Do you think it is relevant to implement unit tests for fragment and activities ?

makingthematrix commented:
Please rebase this PR on top of `develop`

gizemb commented on app/src/main/res-v2/layout/fragment_welcome.xml:
```diff
@@ -0,0 +1,135 @@
+    android:background="@color/athens_light_gray">
+
+    <androidx.constraintlayout.widget.Guideline
+        android:id="@+id/guideline1"
```
those ids are also likely to break uniqueness

mejdoo replied:
I will rename the guidelines based on the related components: 

guideline1 -> guidelineForWelcomeLogoImageView
guideline2 -> guidelineForWelcomeTitleTextView

Does it make sense ?

gizemb replied:
fragmentWelcomeGuidelineLogo
fragmentWelcomeGuidelineTitle would suit better i guess

mejdoo replied:
or probably ? 
guideline1 -> welcomeLogoGuideline
guideline2 -> WelcomeTitleGuideline

gizemb commented on app/src/main/scala/com/waz/zclient/appentry/SSOFragment.scala:
```diff
@@ -79,8 +83,8 @@ trait SSOFragment extends FragmentHelper with DerivedLogTag {
-        title = R.string.app_entry_sso_dialog_title,
-        message = R.string.app_entry_sso_dialog_message,
+        title = R.string.sso_login_dialog_title,
+        message = R.string.sso_login_dialog_message,
```
we should let the QA know if the content of the messages has changed

mejdoo replied:
yes I think so

mejdoo commented:
build again

mejdoo commented:
build again

AndroidBob commented:
Build [2466](https://10.10.124.134/job/AR-build-pipeline/2466/) **failed**.

AndroidBob commented:
Build [2467](https://10.10.124.134/job/AR-build-pipeline/2467/) **failed**.

yamilmedina commented on build-logic/plugins/src/main/kotlin/AndroidCoordinates.kt:
```diff
@@ -1,8 +1,25 @@
+/*
```
duplicated copyright in this file

yamilmedina commented on build-logic/plugins/src/main/kotlin/AndroidApplicationConventionPlugin.kt:
```diff
@@ -1,3 +1,20 @@
+/*
```
duplicated copyright in this file

yamilmedina commented on build-logic/plugins/src/main/kotlin/AndroidLibraryConventionPlugin.kt:
```diff
@@ -1,3 +1,20 @@
+/*
```
duplicated copyright in this file

yamilmedina commented on build-logic/plugins/src/main/kotlin/HiltConventionPlugin.kt:
```diff
@@ -1,10 +1,27 @@
+/*
```
duplicated copyright in this file

yamilmedina commented on build-logic/plugins/src/main/kotlin/com/wire/android/gradle/version/Versionizer.kt:
```diff
@@ -1,7 +1,24 @@
+/*
```
duplicated copyright in this file

AndroidBob commented:
Build [2468](https://10.10.124.134/job/AR-build-pipeline/2468/) **failed**.

AndroidBob commented:
Build [2470](https://10.10.124.134/job/AR-build-pipeline/2470/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-v4.6.0(8550914)(8550914)-staging-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/staging/compat/com.wire.android-v4.6.0(8550914)(8550914)-staging-compat.apk)
